core/bin/custom_genesis_export/src/main.rs

```diff
@@ -0,0 +1,149 @@
+use zksync_types::{url::SensitiveUrl, StorageLog};
+
+#[derive(Debug, Parser)]
+#[command(name = "Custom genesis export tool", author = "Matter Labs")]
+struct Args {
+    /// PostgreSQL connection string for the database to export.
+    #[arg(short, long)]
+    database_url: Option<String>,
+
+    /// Output file path.
+    #[arg(short, long, default_value = "genesis_export.bin")]
+    output_path: PathBuf,
+
+    /// Path to the genesis.yaml
+    #[arg(short, long)]
+    genesis_config_path: PathBuf,
+}
+
+/// The `custom_genesis_export` tool allows exporting storage logs and factory dependencies
+/// from the ZKSync PostgreSQL database in a way that they can be used as a custom genesis state for a new chain.
+///
+/// Inputs:
+///     * `database_url` - URL to the PostgreSQL database.
+///     * `output` - Path to the output file.
+///     * `genesis_config_path` - Path to the `genesis.yaml` configuration file, which will be used to set up a new chain (located in the `file_based` directory).
+///
+/// Given the inputs above, `custom_genesis_export` will perform the following:
+///     * Read storage logs, and factory dependencies; filter out those related to the system context,
+///       and save the remaining data to the output file.
+///     * Calculate the new `genesis_root_hash`, `rollup_last_leaf_index`, and `genesis_commitment`, then update these
+///       in-place in the provided `genesis.yaml`. Additionally, the tool will add a `custom_genesis_state_path` property
+///       pointing to the genesis export.
+///
+/// Note: To calculate the new genesis parameters, the current implementation requires loading all storage logs
+/// into RAM. This is necessary due to the specific sorting and filtering that need to be applied.
+/// For larger states, keep this in mind and ensure you have a machine with sufficient RAM.
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let args = Args::parse();
+
+    let mut out = BufWriter::new(File::create(&args.output_path)?);
+
+    println!(
+        "Export file: {}",
+        args.output_path.canonicalize()?.display(),
+    );
+
+    println!("Connecting to source database...");
+
+    let db_url = args.database_url.or_else(|| std::env::var("DATABASE_URL").ok()).expect("Specify the database connection string in either a CLI argument or in the DATABASE_URL environment variable.");
+    // we need only 1 DB connection at most for data export
+    let connection_pool_builder =
+        ConnectionPool::<Core>::builder(SensitiveUrl::from_str(db_url.as_str())?, 1);
+    let connection_pool = connection_pool_builder.build().await?;
+
+    let mut storage = connection_pool.connection().await?;
+    let mut transaction = storage.start_transaction().await?;
+
+    println!("Connected to source database.");
+
+    let storage_logs = transaction
+        .custom_genesis_export_dal()
+        .get_storage_logs()
+        .await?;
+    let storage_logs_count = storage_logs.len() as u32;
```

Comment:
I hope it will u32, but just in case better to double check as it allows only for ~4bln entries (you use u64 for bytecodes' lengths, so that's why I am double checking)

Reply:
IMO 4bn storage logs should be good enough for genesis state. Even if each storage log represents 1 byte of actual data - that would be an equivalent of 4 Tb state.

Reply:
wdym, 4bn storage is 4gb